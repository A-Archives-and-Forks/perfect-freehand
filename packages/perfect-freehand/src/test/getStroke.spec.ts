import { getStroke } from '../getStroke'
import inputs from './inputs.json'

const { onePoint, twoPoints, twoEqualPoints, manyPoints, withDuplicates } =
  inputs

describe('getStroke', () => {
  for (const [key, value] of Object.entries(inputs)) {
    it(`runs ${key} without generating NaN values`, () => {
      expect(
        getStroke(value).find((t) => JSON.stringify(t).includes('null'))
      ).toBeUndefined()
    })
  }

  it('gets stroke from a line with no points', () => {
    expect(getStroke([])).toMatchSnapshot('get-stroke-no-points')
  })

  it('gets stroke from a line with a single point', () => {
    expect(getStroke(onePoint)).toMatchSnapshot('get-stroke-one-point')
  })

  it('gets stroke from a line with two points', () => {
    expect(getStroke(twoPoints)).toMatchSnapshot('get-stroke-two-points')
  })

  it('gets stroke from a line with two equal points', () => {
    expect(getStroke(twoEqualPoints)).toMatchSnapshot(
      'get-stroke-two-equal-points'
    )
  })

  it('gets stroke from a line with a many points', () => {
    expect(getStroke(manyPoints)).toMatchSnapshot('get-stroke-many-points')
  })

  it('gets stroke from a line with with duplicates', () => {
    expect(getStroke(withDuplicates)).toMatchSnapshot(
      'get-stroke-with-duplicates'
    )
  })

  it('Caps points', () => {
    expect(getStroke(twoPoints).length > 4).toBeTruthy()
  })

  it('Succeeds on tricky points', () => {
    expect(JSON.stringify(getStroke(manyPoints)).includes('null')).toBeFalsy()
  })

  it('Solves a tricky stroke with only one point.', () => {
    const stroke = getStroke(onePoint, {
      size: 1,
      thinning: 0.6,
      smoothing: 0.5,
      streamline: 0.5,
      simulatePressure: true,
      last: false,
    })

    expect(stroke).toMatchSnapshot()

    expect(Number.isNaN([0][0])).toBe(false)
  })
})
