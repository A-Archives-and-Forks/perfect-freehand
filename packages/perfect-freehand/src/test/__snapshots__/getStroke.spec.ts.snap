// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getStroke Solves a tricky stroke with only one point. 1`] = `
Array [
  Array [
    465.24010772327836,
    286.40713984494636,
  ],
  Array [
    465.2500964112529,
    286.57233579644605,
  ],
  Array [
    465.1821682468463,
    286.7232504768442,
  ],
  Array [
    465.05188572523184,
    286.82530899359375,
  ],
  Array [
    464.8890968649966,
    286.8551295110718,
  ],
  Array [
    464.73109695998255,
    286.8058800816854,
  ],
  Array [
    464.61408414362444,
    286.68884386042345,
  ],
  Array [
    464.5648663152038,
    286.53083410868356,
  ],
  Array [
    464.59471938985735,
    286.3680512158076,
  ],
  Array [
    464.6968039610698,
    286.237789108502,
  ],
  Array [
    464.8477322240825,
    286.16989112838985,
  ],
  Array [
    465.01292617454067,
    286.1799128553547,
  ],
  Array [
    465.15453950702556,
    286.26555828921937,
  ],
]
`;

exports[`getStroke gets stroke from a line with a many points: get-stroke-many-points 1`] = `
Array [
  Array [
    0.5508495821598389,
    -10.853391824710275,
  ],
  Array [
    10.487263712280766,
    -25.70547175940249,
  ],
  Array [
    16.909302387744244,
    -34.79842438528664,
  ],
  Array [
    23.53504451646879,
    -44.502838180176255,
  ],
  Array [
    30.392699457210753,
    -54.98003015785006,
  ],
  Array [
    37.74339990102968,
    -66.54666595477848,
  ],
  Array [
    45.492278070157795,
    -78.55740883476514,
  ],
  Array [
    53.41943624372377,
    -91.07701326875332,
  ],
  Array [
    61.11663417290783,
    -103.13773016048455,
  ],
  Array [
    66.83916275239665,
    -112.8234188226388,
  ],
  Array [
    72.2686181495529,
    -121.89389447146331,
  ],
  Array [
    77.1070920654364,
    -130.19614306631112,
  ],
  Array [
    81.56386517750026,
    -137.7900260266369,
  ],
  Array [
    86.225083882178,
    -144.79556338527095,
  ],
  Array [
    94.22836246806321,
    -157.6734732272444,
  ],
  Array [
    101.85858473906248,
    -170.05569306449985,
  ],
  Array [
    108.23814675691673,
    -179.39482312230325,
  ],
  Array [
    113.11338626008067,
    -186.94785496465755,
  ],
  Array [
    118.62323256145517,
    -195.59381676477452,
  ],
  Array [
    119.93325669186845,
    -197.119222029652,
  ],
  Array [
    121.4856479506413,
    -198.42903103110712,
  ],
  Array [
    123.30639427690807,
    -199.32925381241918,
  ],
  Array [
    125.28967380863872,
    -199.76756938324158,
  ],
  Array [
    127.32021824626513,
    -199.7185028220039,
  ],
  Array [
    129.28001225166767,
    -199.1849058795507,
  ],
  Array [
    131.05515250517186,
    -198.1977912352483,
  ],
  Array [
    132.5424677718394,
    -196.81453003860292,
  ],
  Array [
    133.6555152184904,
    -195.11551749535585,
  ],
  Array [
    134.32960447522913,
    -193.19950029434835,
  ],
  Array [
    134.52555744278206,
    -191.1778374453267,
  ],
  Array [
    134.23198532546817,
    -189.1680280880634,
  ],
  Array [
    133.46595054853915,
    -187.28688243683317,
  ],
  Array [
    132.27197508922214,
    -185.64373276527067,
  ],
  Array [
    135.3238949673716,
    -174.60614425031852,
  ],
  Array [
    134.40796654502861,
    -156.5654560983854,
  ],
  Array [
    131.40265736852416,
    -141.3091951068142,
  ],
  Array [
    128.46533458044902,
    -127.02351557922663,
  ],
  Array [
    124.99589393006602,
    -112.06407928892905,
  ],
  Array [
    121.41174181056502,
    -96.26573617805788,
  ],
  Array [
    117.6604945096032,
    -79.01402032163087,
  ],
  Array [
    113.75485310869797,
    -61.49424948296114,
  ],
  Array [
    109.76414300596984,
    -43.80400620029065,
  ],
  Array [
    106.02491274195988,
    -26.880431802191975,
  ],
  Array [
    103.14655799826791,
    -11.367373620209019,
  ],
  Array [
    100.81906406209607,
    3.2092015092298887,
  ],
  Array [
    98.75189963948858,
    17.140026089225994,
  ],
  Array [
    97.15696861228545,
    30.03644961395648,
  ],
  Array [
    96.25181867630182,
    41.50113153346337,
  ],
  Array [
    95.53681323750074,
    52.05969003852795,
  ],
  Array [
    95.15198424587182,
    61.83366998223893,
  ],
  Array [
    95.13648775167786,
    76.51487184148942,
  ],
  Array [
    95.46900532851154,
    88.48534369509058,
  ],
  Array [
    95.84937355209802,
    97.50366709174861,
  ],
  Array [
    96.31956110411606,
    105.86899720252657,
  ],
  Array [
    96.42201088671735,
    107.41722242825804,
  ],
  Array [
    96.19339873023004,
    109.29002762522434,
  ],
  Array [
    95.52322444102498,
    111.05369688162311,
  ],
  Array [
    94.45043858027876,
    112.60572565740684,
  ],
  Array [
    93.03739151262258,
    113.85590996720519,
  ],
  Array [
    91.36620960002034,
    114.73158904019304,
  ],
  Array [
    89.53402201375263,
    115.18186836750472,
  ],
  Array [
    87.64731558241832,
    115.1805776934806,
  ],
  Array [
    85.81574577245304,
    114.7277920321556,
  ],
  Array [
    84.14576350726794,
    113.84982730744,
  ],
  Array [
    82.73442823457935,
    112.59771087038521,
  ],
  Array [
    81.66376682674725,
    111.04421578714333,
  ],
  Array [
    80.99600617509962,
    109.27963126494303,
  ],
  Array [
    80.76995656008484,
    107.40651503912324,
  ],
  Array [
    78.74460035174697,
    99.51102050737828,
  ],
  Array [
    74.36127774786802,
    85.64352758410783,
  ],
  Array [
    71.34682583138213,
    77.14068673706913,
  ],
  Array [
    67.91963640005311,
    66.68641534776448,
  ],
  Array [
    63.745203196747696,
    55.17644639601926,
  ],
  Array [
    59.13877102546526,
    43.40541226397938,
  ],
  Array [
    54.34836607737416,
    31.80340092980233,
  ],
  Array [
    49.26159908136444,
    20.164206653849508,
  ],
  Array [
    44.32938615038735,
    8.518469858959886,
  ],
  Array [
    39.688381992915126,
    -2.2387370265287476,
  ],
  Array [
    35.26372536257266,
    -11.926273297805608,
  ],
  Array [
    30.959687487414065,
    -21.235473627045142,
  ],
  Array [
    26.985316659740292,
    -29.791475184727044,
  ],
  Array [
    20.67178511289104,
    -42.525503240030126,
  ],
  Array [
    15.283350617750273,
    -52.48757162017931,
  ],
  Array [
    10.72283116470662,
    -59.405717004495735,
  ],
  Array [
    5.9598168788343555,
    -66.80556996935158,
  ],
  Array [
    5.59207867157224,
    -67.00182999529552,
  ],
  Array [
    4.654482913325532,
    -69.46673207993375,
  ],
  Array [
    4.55304158818787,
    -72.10198116688747,
  ],
  Array [
    5.298375629679078,
    -74.63166598111444,
  ],
  Array [
    6.812448366905947,
    -76.79092785479313,
  ],
  Array [
    8.936735984604436,
    -78.35369149975963,
  ],
  Array [
    11.448825008444048,
    -79.1563351498564,
  ],
  Array [
    14.08569905450167,
    -79.11482180304097,
  ],
  Array [
    16.571276714087922,
    -78.23349791768335,
  ],
  Array [
    18.645317350115345,
    -76.60463833722741,
  ],
  Array [
    20.090668360498313,
    -74.39878508972407,
  ],
  Array [
    20.756001112914113,
    -71.84689159247563,
  ],
  Array [
    20.57165509255403,
    -69.21614176827845,
  ],
  Array [
    19.556931376412376,
    -66.78197581817648,
  ],
  Array [
    17.818071805430236,
    -64.79925155970282,
  ],
  Array [
    15.537135435703782,
    -63.47556074610803,
  ],
  Array [
    12.952936907252115,
    -62.94949415499593,
  ],
  Array [
    10.336042488166267,
    -63.27613109732403,
  ],
  Array [
    7.960441714917579,
    -64.42127259900681,
  ],
  Array [
    6.074860615562874,
    -66.26502204253009,
  ],
  Array [
    4.876720028949669,
    -68.61433837445439,
  ],
  Array [
    4.491465590774759,
    -71.22324755466796,
  ],
  Array [
    4.959433542412354,
    -73.81859610331729,
  ],
  Array [
    6.231627515794456,
    -76.12865034225334,
  ],
  Array [
    8.174848466081368,
    -77.91154698286437,
  ],
  Array [
    10.585640646817536,
    -78.98061627623943,
  ],
  Array [
    13.21159348029254,
    -79.22392638515527,
  ],
  Array [
    15.777769011657346,
    -78.61600267698074,
  ],
  Array [
    18.015487985373497,
    -77.22049492443801,
  ],
  Array [
    19.69046063419247,
    -75.18351315759658,
  ],
  Array [
    23.728589484320054,
    -67.79036491954464,
  ],
  Array [
    28.259019572960376,
    -59.554176394278954,
  ],
  Array [
    33.349491156408774,
    -48.421432621521866,
  ],
  Array [
    38.972942605456026,
    -35.143124095219285,
  ],
  Array [
    42.668762045200374,
    -26.517991096584762,
  ],
  Array [
    46.80984959023685,
    -17.107973593541775,
  ],
  Array [
    51.1689611775994,
    -7.2230772128426555,
  ],
  Array [
    55.84494610804276,
    3.479868234822921,
  ],
  Array [
    60.95718818232176,
    14.79040558671775,
  ],
  Array [
    66.57221785429451,
    26.44251132310081,
  ],
  Array [
    71.92161014209849,
    37.96133935187074,
  ],
  Array [
    77.32701691303012,
    49.698429447548904,
  ],
  Array [
    82.5462485833856,
    61.47869836729639,
  ],
  Array [
    86.42400534463177,
    71.5437013469928,
  ],
  Array [
    89.94324921389257,
    80.64128835799266,
  ],
  Array [
    94.50827564342275,
    95.23186430244567,
  ],
  Array [
    96.42201134382653,
    107.41643982549144,
  ],
  Array [
    96.19358646900167,
    109.28926787430932,
  ],
  Array [
    95.52358855007283,
    111.05300413931856,
  ],
  Array [
    94.45095789756785,
    112.60514018592802,
  ],
  Array [
    93.03803585540767,
    113.855465794182,
  ],
  Array [
    91.3669415190685,
    114.73131198098243,
  ],
  Array [
    89.5347989698944,
    115.18177452480103,
  ],
  Array [
    87.6480924189262,
    115.18067252142619,
  ],
  Array [
    85.81647733955272,
    114.72807001934582,
  ],
  Array [
    84.1464072862452,
    113.85027229724626,
  ],
  Array [
    82.7349468089698,
    112.5982969999791,
  ],
  Array [
    81.66413005698293,
    111.0449089906453,
  ],
  Array [
    80.99619295022218,
    109.28039125333297,
  ],
  Array [
    80.76995602471538,
    107.40729764184022,
  ],
  Array [
    81.05286020607011,
    103.10263176921475,
  ],
  Array [
    81.61400404437686,
    93.24889915299345,
  ],
  Array [
    82.21483023048896,
    82.44537131294317,
  ],
  Array [
    83.18846311016978,
    69.19457134140492,
  ],
  Array [
    83.9382435674237,
    61.054440839366976,
  ],
  Array [
    84.92894150491934,
    51.03422768160232,
  ],
  Array [
    86.25723284578287,
    40.458121146605315,
  ],
  Array [
    87.77336569598853,
    28.72105219278804,
  ],
  Array [
    89.80850430436428,
    15.7259250456917,
  ],
  Array [
    92.14854753691519,
    1.7841213992608547,
  ],
  Array [
    94.62857099106155,
    -12.889329060090944,
  ],
  Array [
    97.34902062800934,
    -28.711072963776186,
  ],
  Array [
    100.76920994881625,
    -45.71534016780618,
  ],
  Array [
    104.30159977698617,
    -63.42957404419856,
  ],
  Array [
    107.56983061732264,
    -80.95121388790268,
  ],
  Array [
    110.46673339753966,
    -98.32927554739545,
  ],
  Array [
    112.88571814320935,
    -114.3715607652393,
  ],
  Array [
    115.08569640625693,
    -129.16649477207864,
  ],
  Array [
    116.69461187584118,
    -143.26747028513412,
  ],
  Array [
    117.62272140476864,
    -157.55608277807076,
  ],
  Array [
    117.82643868218716,
    -174.5510396429664,
  ],
  Array [
    118.10962295627576,
    -188.5035185962929,
  ],
  Array [
    119.93268297910122,
    -197.11860503635728,
  ],
  Array [
    121.48494324921218,
    -198.42856927038895,
  ],
  Array [
    123.30559954409723,
    -199.32897412183223,
  ],
  Array [
    125.2888352343545,
    -199.7674880184159,
  ],
  Array [
    127.31938456848428,
    -199.71862451186698,
  ],
  Array [
    129.27923192378202,
    -199.18522355148195,
  ],
  Array [
    131.05447087987474,
    -198.19828642614019,
  ],
  Array [
    132.54192446522515,
    -196.81517396793754,
  ],
  Array [
    133.65514180756495,
    -195.11627273793,
  ],
  Array [
    134.32942266265303,
    -193.20032295542816,
  ],
  Array [
    134.52557779551077,
    -191.17867971181153,
  ],
  Array [
    134.23220666060013,
    -189.16884100738562,
  ],
  Array [
    133.4663600020661,
    -187.28761876208353,
  ],
  Array [
    132.27254886368584,
    -185.64434970119106,
  ],
  Array [
    131.22145893643386,
    -184.8843103079631,
  ],
  Array [
    125.96085426728841,
    -177.72755591680487,
  ],
  Array [
    120.40768139988137,
    -170.33331984830028,
  ],
  Array [
    113.85502001862912,
    -162.01926761239986,
  ],
  Array [
    105.74348271771476,
    -150.20198650864697,
  ],
  Array [
    101.20690262407587,
    -143.5511512200893,
  ],
  Array [
    92.07468191872637,
    -131.209216754576,
  ],
  Array [
    87.23161825975811,
    -123.94621473388507,
  ],
  Array [
    81.93061987199184,
    -115.82309404132988,
  ],
  Array [
    76.16395710517682,
    -107.07436211447894,
  ],
  Array [
    70.0420762822013,
    -97.54876738794232,
  ],
  Array [
    62.63357758405237,
    -85.48605106149196,
  ],
  Array [
    55.13315939141299,
    -72.58027978664668,
  ],
  Array [
    47.87115985980247,
    -60.379072960800826,
  ],
  Array [
    40.83369321196069,
    -48.31898984449831,
  ],
  Array [
    34.05286881600848,
    -37.418462815555,
  ],
  Array [
    27.373075439879276,
    -27.36912993802174,
  ],
  Array [
    21.11680350218738,
    -18.605627668028017,
  ],
  Array [
    10.13646067811521,
    -3.556971148710227,
  ],
  Array [
    3.578319399732546,
    4.769725217170889,
  ],
  Array [
    2.346220729156806,
    5.538199546941077,
  ],
  Array [
    0.9660131906133298,
    5.989472280549046,
  ],
  Array [
    -0.4820854889089605,
    6.097315425514543,
  ],
  Array [
    -1.913911746117605,
    5.855461133185184,
  ],
  Array [
    -3.2462477716695677,
    5.2779659863946256,
  ],
  Array [
    -4.40165813316023,
    4.398394030207829,
  ],
  Array [
    -5.312990326093031,
    3.2678660280448844,
  ],
  Array [
    -5.927277680595956,
    1.952088320509132,
  ],
  Array [
    -6.208817786763128,
    0.5275339697840966,
  ],
  Array [
    -6.141247520396007,
    -0.9230018583336785,
  ],
  Array [
    -5.728494068550624,
    -2.3152139533608587,
  ],
  Array [
    -4.994546681241372,
    -3.568186882921655,
  ],
]
`;

exports[`getStroke gets stroke from a line with a single point: get-stroke-one-point 1`] = `
Array [
  Array [
    470.6756708860476,
    284.71344160555435,
  ],
  Array [
    470.8501335943805,
    287.5987587941242,
  ],
  Array [
    469.66369834235906,
    290.2346388953459,
  ],
  Array [
    467.38818011342,
    292.0171958523662,
  ],
  Array [
    464.54490524935505,
    292.53804185900543,
  ],
  Array [
    461.7852744134242,
    291.6778499242401,
  ],
  Array [
    459.74152523400375,
    289.6336919541033,
  ],
  Array [
    458.8818852426058,
    286.8738891349793,
  ],
  Array [
    459.4032998937971,
    284.0307184969805,
  ],
  Array [
    461.18631191880905,
    281.75555682494087,
  ],
  Array [
    463.82242925436196,
    280.56964877266483,
  ],
  Array [
    466.70771149268404,
    280.7446885409423,
  ],
  Array [
    469.1811339866371,
    282.2405741253579,
  ],
]
`;

exports[`getStroke gets stroke from a line with no points: get-stroke-no-points 1`] = `Array []`;

exports[`getStroke gets stroke from a line with two equal points: get-stroke-two-equal-points 1`] = `
Array [
  Array [
    6.728352660193719,
    -0.7849301948399909,
  ],
  Array [
    6.901686161036018,
    2.081711817738876,
  ],
  Array [
    5.722930094936034,
    4.700531221101572,
  ],
  Array [
    3.462140126572477,
    6.471550602628259,
  ],
  Array [
    0.6372683201289022,
    6.989025440622029,
  ],
  Array [
    -2.1045008430072114,
    6.134401086374877,
  ],
  Array [
    -4.13502188385132,
    4.103473900706759,
  ],
  Array [
    -4.98909786717701,
    1.361533867536321,
  ],
  Array [
    -4.4710580651752165,
    -1.4632343874429035,
  ],
  Array [
    -2.6995865610782586,
    -3.7236701067167184,
  ],
  Array [
    -0.08053145888030677,
    -4.9019023853644,
  ],
  Array [
    2.786075829665783,
    -4.727995559590081,
  ],
  Array [
    5.243489130398231,
    -3.241792074134711,
  ],
]
`;

exports[`getStroke gets stroke from a line with two points: get-stroke-two-points 1`] = `
Array [
  Array [
    2.9164089923107177,
    196.28210564488668,
  ],
  Array [
    5.4556846089458855,
    193.41598924464512,
  ],
  Array [
    9.03607309764555,
    192.058284509823,
  ],
  Array [
    12.837299406657174,
    192.52004464739375,
  ],
  Array [
    15.988494043134171,
    194.69547937176728,
  ],
  Array [
    17.76771123172934,
    198.08619169704335,
  ],
  Array [
    17.767328314717073,
    201.9153618069165,
  ],
  Array [
    15.987433019245714,
    205.30571822094296,
  ],
  Array [
    12.835803358850638,
    207.48052266288468,
  ],
  Array [
    9.034484773836642,
    207.9415225459635,
  ],
  Array [
    5.454367897689907,
    206.58310176060252,
  ],
  Array [
    2.915665555116475,
    203.71647756256334,
  ],
  Array [
    2.000000159999999,
    199.9984000000107,
  ],
]
`;

exports[`getStroke gets stroke from a line with with duplicates: get-stroke-with-duplicates 1`] = `
Array [
  Array [
    10.168344244742169,
    1.3316557552578319,
  ],
  Array [
    65.80475920006808,
    57.417924979619414,
  ],
  Array [
    87.50286525240143,
    79.86677552396576,
  ],
  Array [
    96.87876278474482,
    89.25333454521123,
  ],
  Array [
    102.96676306431297,
    94.52834701591033,
  ],
  Array [
    103.75529212050765,
    95.18012966358725,
  ],
  Array [
    104.65681228214318,
    96.33084424767203,
  ],
  Array [
    105.25674162717746,
    97.66387366851295,
  ],
  Array [
    105.52021223388954,
    99.10174219353863,
  ],
  Array [
    105.4319111782382,
    100.5608808365166,
  ],
  Array [
    104.99697052172671,
    101.95648439095966,
  ],
  Array [
    104.2406690357471,
    103.20744031271143,
  ],
  Array [
    103.2069629981992,
    104.24104298482376,
  ],
  Array [
    101.95593145255374,
    104.99721937142988,
  ],
  Array [
    100.56028441102313,
    105.43202046541147,
  ],
  Array [
    99.1011369452353,
    105.52017560675725,
  ],
  Array [
    97.66329477445959,
    105.25656121451026,
  ],
  Array [
    96.33032535321821,
    104.65649856953375,
  ],
  Array [
    95.17970092690302,
    103.75486334094761,
  ],
  Array [
    38.43703385226901,
    46.11052040597133,
  ],
  Array [
    14.129612003024915,
    21.803098556727228,
  ],
  Array [
    13.105350408218428,
    20.377128928032562,
  ],
  Array [
    12.590037306655162,
    18.698750661993472,
  ],
  Array [
    12.637626116098236,
    16.943690406113085,
  ],
  Array [
    13.243134275636738,
    15.295703432472994,
  ],
  Array [
    14.343164921128356,
    13.927334412239624,
  ],
  Array [
    15.822544553610225,
    12.981851943045752,
  ],
  Array [
    17.526381735043202,
    12.558248288999108,
  ],
  Array [
    19.276284262761916,
    12.700874864498882,
  ],
  Array [
    20.88903688018481,
    13.394798629440485,
  ],
  Array [
    22.19578396254237,
    14.567365582456453,
  ],
  Array [
    23.059708745364023,
    16.095807654038392,
  ],
  Array [
    23.39035807583435,
    17.820096555822158,
  ],
  Array [
    23.153112881770163,
    19.55969878462292,
  ],
  Array [
    22.37281279761663,
    21.132477527874848,
  ],
  Array [
    21.131155448163838,
    22.373762435986208,
  ],
  Array [
    19.558142685665217,
    23.153590651410333,
  ],
  Array [
    17.818469361589404,
    23.390313954137667,
  ],
  Array [
    16.0942797321963,
    23.05914735188378,
  ],
  Array [
    14.566096906825212,
    22.194764075324144,
  ],
  Array [
    13.39392203069359,
    20.88766528168958,
  ],
  Array [
    12.700482122756526,
    19.274704559714188,
  ],
  Array [
    12.558380524425392,
    17.524759322769086,
  ],
  Array [
    12.982495310556633,
    15.821049299103091,
  ],
  Array [
    13.928421551086881,
    14.341953377929807,
  ],
  Array [
    15.29712051893234,
    13.242333292639481,
  ],
  Array [
    16.94528907085815,
    12.637319556433523,
  ],
  Array [
    18.70036352440344,
    12.590257267200807,
  ],
  Array [
    20.37858712098736,
    13.106073859047239,
  ],
  Array [
    21.804249407039563,
    14.130763198644143,
  ],
  Array [
    46.11052040597133,
    38.43703385226901,
  ],
  Array [
    103.75486334094761,
    95.17970092690302,
  ],
  Array [
    104.65649856953375,
    96.33032535321821,
  ],
  Array [
    105.25656121451026,
    97.66329477445959,
  ],
  Array [
    105.52017560675725,
    99.1011369452353,
  ],
  Array [
    105.43202046541147,
    100.56028441102313,
  ],
  Array [
    104.99721937142988,
    101.95593145255374,
  ],
  Array [
    104.24104298482376,
    103.2069629981992,
  ],
  Array [
    103.20744031271143,
    104.2406690357471,
  ],
  Array [
    101.95648439095966,
    104.99697052172671,
  ],
  Array [
    100.5608808365166,
    105.4319111782382,
  ],
  Array [
    99.10174219353863,
    105.52021223388954,
  ],
  Array [
    97.66387366851295,
    105.25674162717746,
  ],
  Array [
    96.33084424767203,
    104.65681228214318,
  ],
  Array [
    95.18012966358725,
    103.75529212050765,
  ],
  Array [
    94.52834701591033,
    102.96676306431297,
  ],
  Array [
    89.25333454521123,
    96.87876278474482,
  ],
  Array [
    79.86677552396576,
    87.50286525240143,
  ],
  Array [
    57.417924979619414,
    65.80475920006808,
  ],
  Array [
    1.3316557552578319,
    10.168344244742169,
  ],
  Array [
    -5.347359044798066,
    3.2325350695868424,
  ],
  Array [
    -5.965585085925712,
    1.8588508043071195,
  ],
  Array [
    -6.237091041567539,
    0.3771300429433353,
  ],
  Array [
    -6.146096972824996,
    -1.1265095342356943,
  ],
  Array [
    -5.697891459248734,
    -2.564676324356122,
  ],
  Array [
    -4.918524226346653,
    -3.853784006004171,
  ],
  Array [
    -3.853292134313433,
    -4.918909580153992,
  ],
  Array [
    -2.5641065223877635,
    -5.698147898391285,
  ],
  Array [
    -1.125894918906889,
    -6.146209593047747,
  ],
  Array [
    0.3777537501608017,
    -6.237053297377853,
  ],
  Array [
    1.8594473535204608,
    -5.9653991710176655,
  ],
  Array [
    3.2330697893277534,
    -5.347035764554851,
  ],
  Array [
    4.418786057074182,
    -4.417902388226714,
  ],
]
`;
